src/service/activity-impl.js

```diff
@@ -0,0 +1,292 @@
+    this.ignoreInactive_ = false;
+
+    /** @private @const {!ActivityHistory} */
+    this.activityHistory_ = new ActivityHistory();
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    this.viewer_.whenFirstVisible().then(this.start_.bind(this));
+  }
+
+  start_() {
+    /** @private @const {number} */
+    this.startTime_ = (new Date()).getTime();
+    // record an activity since this is when the page became visible
+    this.handleActivity_();
+    this.setUpActivityListeners_();
+  }
+
+  getTimeSinceStart_() {
+    const timeSinceStart = (new Date()).getTime() - this.startTime_;
+    // Ensure that a negative time is never returned. This may cause loss of
+    // data if there is a time change during the session but it will decrease
+    // the likelyhood of errors in that situation.
+    return (timeSinceStart > 0 ? timeSinceStart : 0);
+  }
+
+  /**
+   * Return to a state where neither activities or inactivity events are
+   * ignored when that event type is fired.
+   */
+  stopIgnore_() {
+    this.ignoreActivity_ = false;
+    this.ignoreInactive_ = false;
+  }
+
+  setUpActivityListeners_() {
+    for (let i = 0; i < ACTIVE_EVENT_TYPES.length; i++) {
+      this.unlistenFuncs_.push(listen(this.win_.document.documentElement,
+        ACTIVE_EVENT_TYPES[i], this.boundHandleActivity_));
+    }
+
+    this.viewer_.onVisibilityChanged(this.boundHandleVisibilityChange_);
+  }
+
+  handleActivity_() {
+    if (this.ignoreActivity_) {
+      return;
+    }
+    this.ignoreActivity_ = true;
+    this.ignoreInactive_ = false;
+
+    this.handleActivityEvent_(ActivityEventType.ACTIVE);
+  }
+
+  handleInactive_() {
+    if (this.ignoreInactive_) {
+      return;
+    }
+    this.ignoreInactive_ = true;
+    this.ignoreActivity_ = false;
+
+    this.handleActivityEvent_(ActivityEventType.INACTIVE);
+  }
+
+  /**
+   * @param {ActivityEventType}
+   */
+  handleActivityEvent_(type) {
+    const timeSinceStart = this.getTimeSinceStart_();
+    const secondKey = Math.floor(timeSinceStart / 1000);
+    const timeToWait = 1000 - (timeSinceStart % 1000);
+
+    // stop ignoring activity at the start of the next activity bucket
+    setTimeout(this.boundStopIgnore_, timeToWait);
+
+    this.activityHistory_.push(/** @type {ActivityEventDef} */{
+      type: type,
+      time: secondKey
+    });
+  }
+
+  handleVisibilityChange_() {
+    if (this.viewer_.isVisible()) {
+      this.handleActivity_();
+    } else {
+      this.handleInactive_();
+    }
+  }
+
+  /**
+   * Remove all listeners associated with this Activity instance.
+   */
+  unlisten_() {
+    for (let i = 0; i < this.unlistenFuncs_.length; i++) {
+      const unlistenFunc = this.unlistenFuncs_[i];
+      // TODO(britice): Due to eslint typechecking, this check may not be
+      // necessary.
+      if (typeof unlistenFunc === 'function') {
+        unlistenFunc();
+      }
+    }
+    this.unlistenFuncs_ = [];
+  }
+
+  cleanup_() {
```

Comment:
It's not being called at this point. I saw `cleanup_` functionality in other components and decided to set this up here as well. Since the activity tracking is intended to be available until the user leaves the page, it didn't seem necessary to ensure it was called anywhere.

I'll add the annotations for all the methods.